server/src/services/documentSymbol/onDocumentSymbol.ts

```diff
@@ -0,0 +1,140 @@
+import { analyze } from "@nomicfoundation/solidity-analyzer";
+import semver from "semver";
+import _ from "lodash";
+import { Language } from "@nomicfoundation/slang/language";
+import { ProductionKind } from "@nomicfoundation/slang/kinds";
+import { Cursor } from "@nomicfoundation/slang/cursor";
+import { RuleNode } from "@nomicfoundation/slang/cst";
+import { ServerState } from "../../types";
+import { SlangNode } from "../../parser/slangHelpers";
+import { SymbolTreeBuilder } from "./SymbolTreeBuilder";
+import { StructDefinition } from "./visitors/StructDefinition";
+import { StructMember } from "./visitors/StructMember";
+import { InterfaceDefinition } from "./visitors/InterfaceDefinition";
+import { FunctionDefinition } from "./visitors/FunctionDefinition";
+import { ContractDefinition } from "./visitors/ContractDefinition";
+import { EventDefinition } from "./visitors/EventDefinition";
+import { StateVariableDeclaration } from "./visitors/StateVariableDeclaration";
+import { VariableDeclaration } from "./visitors/VariableDeclaration";
+import { ConstantDefinition } from "./visitors/ConstantDefinition";
+import { ConstructorDefinition } from "./visitors/ConstructorDefinition";
+import { EnumDefinition } from "./visitors/EnumDefinition";
+import { ErrorDefinition } from "./visitors/ErrorDefinition";
+import { FallbackFunctionDefinition } from "./visitors/FallbackFunctionDefinition";
+import { LibraryDefinition } from "./visitors/LibraryDefinition";
+import { ModifierDefinition } from "./visitors/ModifierDefinition";
+import { ReceiveFunctionDefinition } from "./visitors/ReceiveFunctionDefinition";
+import { UserDefinedValueTypeDefinition } from "./visitors/UserDefinedValueTypeDefinition";
+import { SymbolVisitor } from "./SymbolVisitor";
+import { YulFunctionDefinition } from "./visitors/YulFunctionDefinition";
+import { UnnamedFunctionDefinition } from "./visitors/UnnamedFunctionDefinition";
+
+export function onDocumentSymbol(serverState: ServerState) {
+  return async (
+    params: DocumentSymbolParams
+  ): Promise<DocumentSymbol[] | SymbolInformation[] | null> => {
+    const { telemetry } = serverState;
+    return telemetry.trackTimingSync("onDocumentSymbol", (transaction) => {
+      const { uri } = params.textDocument;
+
+      // Find the file in the documents collection
+      const document = serverState.documents.get(uri);
+
+      if (document === undefined) {
+        throw new Error(`${uri} not found in documents`);
+      }
+
+      const text = document.getText();
+
+      // Get the document's solidity version
+      let span = transaction.startChild({ op: "solidity-analyzer" });
+      const { versionPragmas } = analyze(text);
+      span.finish();
+
+      const versions = Language.supportedVersions();
+      versionPragmas.push(
+        `>= ${versions[0]}`,
+        `<= ${versions[versions.length - 1]}`
+      );
+
+      const solcVersion = semver.maxSatisfying(
```

Comment:
left over:

```suggestion
      const slangVersion = semver.maxSatisfying(
```

Reply:
Also, I suggest reusing `versions` created above instead of fetching it two more times.